### PR TITLE
[ASR] Set default sampling parameters for MOSS-Transcribe-Diarize

### PR DESCRIPTION
**Pull request by CloudRipple**
## Motivation
MOSS-Transcribe-Diarize should use the model’s intended sampling defaults instead of inheriting route-level OpenAI/client defaults. In particular, omitted sampling fields should resolve to `temperature=0.8`, `top_p=0.95`, and `top_k=50`, while explicit user-provided values should still be preserved.

The implementation follows the existing TTS pattern by recording explicit generation params separately, so implicit request defaults from `chat`, `/generate`, or `transcriptions` routes do not accidentally override model-specific defaults.

## Modifications

**Core logic for explicit/implicit parameter handling:**
* Added `_explicit_generation_fields` and `_sampling_param` helpers in `request_builders.py` to determine which sampling parameters are explicitly set by the user and to apply model-specific defaults or ignore implicit client defaults accordingly. [[1]](diffhunk://#diff-0e9cd1b235a85c0c7a585dd898c148d7d2d7c8971eca5fee765a700874a53bf4R120-R147) [[2]](diffhunk://#diff-0e9cd1b235a85c0c7a585dd898c148d7d2d7c8971eca5fee765a700874a53bf4R33-R40)
* Updated the MOSS-Transcribe-Diarize request builder to use these helpers, ensuring only explicit sampling parameters are preserved and that model defaults are applied when appropriate. [[1]](diffhunk://#diff-0e9cd1b235a85c0c7a585dd898c148d7d2d7c8971eca5fee765a700874a53bf4R242-R243) [[2]](diffhunk://#diff-0e9cd1b235a85c0c7a585dd898c148d7d2d7c8971eca5fee765a700874a53bf4L251-R299) [[3]](diffhunk://#diff-0e9cd1b235a85c0c7a585dd898c148d7d2d7c8971eca5fee765a700874a53bf4R329-R330)

**OpenAI API integration:**
* Added `_explicit_generation_params` and `_add_explicit_generation_params` helpers to annotate request metadata with a list of explicit sampling parameters, and updated both chat and rollout request builders to use this mechanism. [[1]](diffhunk://#diff-c40f01fbad12030d6a574e4a3aec3ee4566e9f1c8b358ad686b561213e5aa3dfR846-R870) [[2]](diffhunk://#diff-c40f01fbad12030d6a574e4a3aec3ee4566e9f1c8b358ad686b561213e5aa3dfR940-R943) [[3]](diffhunk://#diff-c40f01fbad12030d6a574e4a3aec3ee4566e9f1c8b358ad686b561213e5aa3dfR1059-R1063) [[4]](diffhunk://#diff-c40f01fbad12030d6a574e4a3aec3ee4566e9f1c8b358ad686b561213e5aa3dfL1046-R1079)
* Modified `build_transcription_generate_request` to record explicit sampling parameters in metadata and use the new mechanism for parameter handling. [[1]](diffhunk://#diff-c40f01fbad12030d6a574e4a3aec3ee4566e9f1c8b358ad686b561213e5aa3dfR1643-R1657) [[2]](diffhunk://#diff-c40f01fbad12030d6a574e4a3aec3ee4566e9f1c8b358ad686b561213e5aa3dfR1666-R1670)

**Testing and validation:**
* Added and updated unit tests in `test_request_builders.py`, `test_generate_rollout.py`, and `test_openai_api.py` to verify that explicit and implicit parameter handling works as intended, including edge cases for default values and metadata propagation. [[1]](diffhunk://#diff-315358e405d402e9d8d4310f2c88dc3a3dc2de07b76b64e93bcfe0196d8b1959R138-R228) [[2]](diffhunk://#diff-fdfb1eae09c4cd630087f5e31bc358aa06881d74d5979c2b4b6ed0e6e90f33afR416-R470) [[3]](diffhunk://#diff-d2c9f3a6ce3c757dbadeae79339f971f9c2f70bd88fad1ac6ad1a5664346f010R589-R640) [[4]](diffhunk://#diff-d2c9f3a6ce3c757dbadeae79339f971f9c2f70bd88fad1ac6ad1a5664346f010L837-R894) [[5]](diffhunk://#diff-d2c9f3a6ce3c757dbadeae79339f971f9c2f70bd88fad1ac6ad1a5664346f010L857-R913) [[6]](diffhunk://#diff-d2c9f3a6ce3c757dbadeae79339f971f9c2f70bd88fad1ac6ad1a5664346f010L875-R934)

**Constants and test utilities:**
* Introduced `DEFAULT_TEMPERATURE`, `DEFAULT_TOP_P`, and `DEFAULT_TOP_K` constants for MOSS sampling defaults and updated test imports and payload construction helpers accordingly. [[1]](diffhunk://#diff-0e9cd1b235a85c0c7a585dd898c148d7d2d7c8971eca5fee765a700874a53bf4R33-R40) [[2]](diffhunk://#diff-315358e405d402e9d8d4310f2c88dc3a3dc2de07b76b64e93bcfe0196d8b1959R14-R16) [[3]](diffhunk://#diff-315358e405d402e9d8d4310f2c88dc3a3dc2de07b76b64e93bcfe0196d8b1959L76-R92)

**Minor refactors:**
* Simplified the `_rollout_sampling_to_client` function for clarity. [[1]](diffhunk://#diff-c40f01fbad12030d6a574e4a3aec3ee4566e9f1c8b358ad686b561213e5aa3dfL987-R1016) [[2]](diffhunk://#diff-c40f01fbad12030d6a574e4a3aec3ee4566e9f1c8b358ad686b561213e5aa3dfL998-R1028)

These changes make the model's behavior more predictable and robust by ensuring that only user-intended sampling settings affect generation, and that implicit client defaults do not unintentionally override model-specific defaults.
#### Runs

| Metric | Run 1 | Run 2 | Run 3 | Run 4 | Run 5 | Run 6 | Run 7 |
| --- | ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| cer_no_spk | 5.84 | 5.97 | 5.84 | 5.87 | 5.89 | 5.74 | 5.88 |
| cer_no_spk_below_50_corpus | 5.01 | 5.18 | 5.00 | 5.00 | 5.04 | 4.93 | 4.95 |
| n_above_50_pct_cer | 30 | 28 | 29 | 27 | 28 | 28 | 32 |
| cp_cer | 12.83 | 13.53 | 12.91 | 13.54 | 13.87 | 12.86 | 12.74 |
| delta_cer | 6.99 | 7.56 | 7.07 | 7.67 | 7.98 | 7.12 | 6.86 |
| speaker_timestamp_der | 21.65 | 22.02 | 21.77 | 22.15 | 21.99 | 21.81 | 21.70 |

#### Summary

| Metric | Avg | Min | Max |
| --- | ---: | ---: | ---: |
| cer_no_spk | 5.86 | 5.74 | 5.97 |
| cer_no_spk_below_50_corpus | 5.02 | 4.93 | 5.18 |
| n_above_50_pct_cer | 28.86 | 27.00 | 32.00 |
| cp_cer | 13.18 | 12.74 | 13.87 |
| delta_cer | 7.32 | 6.86 | 7.98 |
| speaker_timestamp_der | 21.87 | 21.65 | 22.15 |

#### Runs

| Metric | Run 1 | Run 2 | Run 3 | Run 4 | Run 5 |
| --- | ---: | ---: | ---: | ---: | ---: |
| cer_no_spk | 13.93 | 13.62 | 13.62 | 13.62 | 13.38 |
| cp_cer | 14.56 | 14.36 | 14.36 | 14.32 | 14.33 |
| delta_cer | 0.63 | 0.74 | 0.74 | 0.70 | 0.95 |
| speaker_timestamp_der | 17.62 | 15.64 | 15.64 | 19.02 | 17.65 |

#### Summary

| Metric | Avg | Min | Max |
| --- | ---: | ---: | ---: |
| cer_no_spk | 13.63 | 13.38 | 13.93 |
| cp_cer | 14.39 | 14.32 | 14.56 |
| delta_cer | 0.75 | 0.63 | 0.95 |
| speaker_timestamp_der | 17.11 | 15.64 | 19.02 |

## Checklist

- [x] Format your code according with pre-commit.
- [x] Add unit tests.
- [ ] Update documentation / docstrings / example tutorials as needed.
- [x] Provide throughput / latency benchmark results and accuracy evaluation results as needed.
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.

## CI

CI runs on self-hosted GPU runners and requires a maintainer to add the
`run-ci` label. Once labeled, every subsequent push re-triggers CI as
long as the label remains. Use `/tag-and-rerun-ci higgs` or
`/tag-and-rerun-ci moss` to select a TTS CI model. Draft PRs are skipped even
if labeled.

### CODE REVIEW

File: sglang_omni/models/moss_transcribe_diarize/request_builders.py
Comment (zhaochenyang20) [inline sglang_omni/models/moss_transcribe_diarize/request_builders.py:1]: I verified that every server-side production route (chat / transcription / rollout) now stamps `asr_params` (even an empty `{}`), so for real API traffic `explicit_fields` is never None and the sniffing branch never fires — it only serves hand-built requests that carry `params` but no `asr_params` (test `test_request_builder_ignores_implicit_client_sampling_defaults` is exactly that). Since the primary goal (omit → 0.8/0.95/50) is already fully achieved by the empty-set branch, I recommend **deleting the value-sniffing outright** and collapsing the `None` branch into the empty-set semantics: any field not marked explicit gets the model default. This reduces three states to two and removes the "user wants 1.0 but gets 0.8" silent-bug surface.

File: sglang_omni/models/moss_transcribe_diarize/request_builders.py [hunk 1]
```diff
@@ -109,6 +117,34 @@ def _load_audio(source: Any) -> np.ndarray:
...
+    value = params.get(field)
+    if value is None:
+        return default
+    normalized = cast(value)
+    if explicit_fields is not None:
+        return normalized if field in explicit_fields else default
+    if normalized in _IMPLICIT_SAMPLING_DEFAULTS.get(field, set()):
+        return default
+    return normalized
+
```
Comment (zhaochenyang20) [inline sglang_omni/models/moss_transcribe_diarize/request_builders.py]: Drop `_IMPLICIT_SAMPLING_DEFAULTS` and the tri-state; simplify `_sampling_param` to two branches — any field not marked explicit uses the model default.

```python
def _sampling_param(params, explicit_fields, field, default, cast):
    if field not in explicit_fields:
        return default
    value = params.get(field)
    return default if value is None else cast(value)
```

File: sglang_omni/serve/openai_api.py
Comment (zhaochenyang20) [inline sglang_omni/serve/openai_api.py:1]: `_add_explicit_generation_params` unconditionally injects an `asr_params` key into the metadata of **every** chat and rollout request — even a plain-text Qwen chat or an RL rollout that has nothing to do with ASR. Test `test_converter_preserves_rollout_metadata` even freezes this (`gen.metadata == {"rollout_id": 1, "asr_params": {}}`). The key name (`asr_params`) is misleading on non-ASR requests; the real concept it carries is "which sampling fields the user set explicitly," which is neither ASR- nor TTS-specific. Worse, the same concept is `tts_params["explicit_generation_params"]` on the TTS side and `asr_params["explicit_generation_params"]` here — one concept, two namespaces, held together only by convention.

Carry this concept under a single model-family-neutral key (e.g. `explicit_sampling`, or a top-level `metadata["explicit_generation_params"]`) shared by TTS/ASR/chat/rollout; and only write the key when there actually are explicit fields — don't pollute metadata with an empty set (the reader already tolerates a missing key, see the None branch of `_explicit_generation_fields`).

**Review by zhaochenyang20 — CHANGES_REQUESTED** [review]
Write side: openai_api's `_explicit_generation_params(request)` and speech_service's `_explicit_generation_params(request)` are two identically-named, near-identical module-private functions (near-identical field tuples; speech_service adds `seed`). Read side: the 6-line block `explicit_generation_params = X.get("explicit_generation_params"); if isinstance(...(list, tuple, set)): {str(f) for f in ...}` is copied verbatim across 5 model files. 5 sites × 6 lines = 30 lines of duplicated read side, plus 2 near-duplicate write sides — clearly past the "worth a shared helper" threshold (repo P2 rule: >5 duplicated lines → extract). This PR adds yet another copy — MTD's `_explicit_generation_fields` is the 5th.

Extract a shared module (e.g. `sglang_omni/serve/sampling_defaults.py`, or under `client/`) exposing a pair: `explicit_sampling_fields(request) -> list[str]` (write side, shared by openai_api + speech_service, with a parameterizable field set to accommodate `seed`) and `explicit_fields_from_metadata(metadata) -> set[str]` (read side, shared by all model request builders), plus a single metadata-key constant. The 5 duplicated read sites collapse into one import.

File: sglang_omni/models/moss_transcribe_diarize/request_builders.py [hunk 2]
```diff
@@ -30,6 +30,14 @@
...
 _AUDIO_END = "<|audio_end|>"
 _SPECIAL_TOKEN_RE = re.compile(r"<\|(?:im_start|im_end|endoftext)\|>")
+DEFAULT_TEMPERATURE = 0.8
+DEFAULT_TOP_P = 0.95
+DEFAULT_TOP_K = 50
+_IMPLICIT_SAMPLING_DEFAULTS = {
+    "temperature": {1.0},
+    "top_p": {1.0},
+    "top_k": {-1},
+}
```
Comment (zhaochenyang20) [inline sglang_omni/models/moss_transcribe_diarize/request_builders.py]: Two small over-generalizations. First, `isinstance(explicit_generation_params, (list, tuple, set))` accepts three container types, but the producer (`_add_explicit_generation_params`) only ever writes a `list`; tolerating tuple/set is defensive code for a case that never happens (copied from TTS, so it's "following the existing mistake," not original to this PR). Second, `_IMPLICIT_SAMPLING_DEFAULTS = {"temperature": {1.0}, ...}` uses single-element sets as "the set of implicit defaults," but each field has exactly one implicit default; the set structure implies "there could be several" and never delivers.

**Review by zhaochenyang20 — CHANGES_REQUESTED** [review]

**Comment by CloudRipple** [comment]
moved to #989 and merged
